### PR TITLE
fix DDB performance when Streams are enabled

### PR DESCRIPTION
**Pull request by bentsku**
## Motivation
Following #10411 and other reports from users, we need to investigate the performance issues of DynamoDB when streams are configured for the table (DynamoDB Streams or Kinesis).

This PR will try to explain the different changes done to get better results. After the changes, benchmarks will be provided to show the progress.

Maybe a bit of explanation on how we handle streams right now:

If the table has a stream configured, we will try to fetch the item to be inserted with a `connect_to` call directly to LocalStack (there can also be additional calls for the schema of the table), to check if the item already existed and if it will be updated.
We prepare the record to be sent to DynamoDB streams and/or Kinesis, if they're configured. 
Then, we forward the event to be sent (always via Kinesis under the hood) via a spawn thread for each call to forward, which calls `PutRecord` for each update.
For `PutItem`, this isn't so bad, as we can only modify one item each time.
But for `BatchWriteItem`, there is a lot of issues: we fetch each individual objects one by one (each time going through the LocalStack Gateway, having the request parsed and going through the full chain), and we then forward each update with a singular call to `PutRecord`. 

## Changes
 - refactor the `ItemFinder` utility class to directly send a call to the underlying DynamoDB-local server, as we don't need to go through the full gateway
 - add a method to the `ItemFinder` to be able to fetch multiple items with a singular call to `BatchGetItem`, making it more efficient. This call also bypass the LocalStack gateway
 - refactor `PutItem` and `DeleteItem` logic to modify the incoming request to add `ReturnValues` if it's not set in the request. This way, we don't have the fetch the previous item: DynamoDB will return it to us almost for free. If the client didn't ask for it, we remove it from the request. We only do this if streams are enabled on the table. This is not done for `UpdateItem`, as the logic is more complicated and you need both the old value and new value, and `ReturnValues` accept different values. Not worth it for now.
 - refactor `BatchWriteItem` to use the new `ItemFinder.find_existing_items` allowing to fetch all modified objects in one call, plus additional logic to map the written items to the retrieved ones (before it was easy as we fetched them individually), but `BatchGetItem` could return less items if they don't exist, and the order is not guaranteed. 
 - refactor the forwarders to DynamoDB streams and Kinesis (both Kinesis under the hood) to use `PutRecords` instead of `PutRecord`, that way, we can batch records that share the same table and stream, lowering the load on the gateway. 

### Scenario 1

This is the scenario provided in the linked GitHub issue (just added a small check if the streams were ready). 
We're testing against 3 setups:
- No stream configured: our baseline
- DDB stream configured
- Kinesis stream configured

We can test with 3 variables:
- `latest` image with no fix
- this PR with the hypercorn server (default)
- this PR with the Twisted server (to verify the speed boost)

What the test looks like: we are starting up 300 threads, waiting on them and concurrently launching 300 `BatchWriteItem` request with 10 items in them. This does an "onslaught" on request. This will not test throughput in itself, but how quick LocalStack can go through those 300 requests. 

**Results**

TLDR; *6x* speed and throughput improvement for `BatchWriteItem` when doing 300 concurrent requests inserting 10 items each.

| / | `latest` | Fix & Hypercorn | Fix & Twisted |
|--------|--------|--------|--------|
| **NoStream** | 0.816 | 0.796 | 0.704 |
| **DDB Stream** | 11.54 | 1.948 | 1.725 |
| **Kinesis** | 11.18 | 2.154 | 2.29 |

As we can see, the baselines are all pretty close, which shows the benchmark is kinda stable.
We can notice around a *6x decrease* in time for the benchmark with the fix. Twisted is not giving much more here, I think because of the shape of the benchmark: just go through all the concurrent requests, only 300, which is not so many, and the pressure must be also on the logic behind. 
We're still around 2x to 3x slower than with no streams at all. I don't think we can go much lower easily with all the additional logic and pressure that happens: we are fetching old items, doing a lot of dict manipulation to format the records, and forwarding them over. We will need to refactor it to streamline it and maybe reduce dict manipulation and such, and simplify the forwarding with some thread workers, or find a way to hook into some kind of transaction log.

I wanted to go further in the testing, to test continuous throughput. 

### Scenario 2

This scenario is the same as the previous DynamoDB performance rework, see the following comment: https://github.com/localstack/localstack/issues/1205#issuecomment-1776138503

We are testing against the same 3 setups, with the same 3 variables. We will just add a check for `PutItem` as well, to see the upgrade.

**Results**
TLDR; *15x* speed and throughput improvement for `BatchWriteItem` when doing 500 requests inserting 25 items each. Now only 2x slower than baseline (but doing a lot more) instead of 30x.

No Stream | `latest` | Fix & Hypercorn | Fix & Twisted
-- | -- | -- | --
**Throughput** (higher is better) |   |   |  
`PutItem` throughput item/s | 333.01 | 313.88 | 473.38
`BatchWriteItem (25)` throughput item/s | 4985.24 | 4943.65 | 6209.03
**Total time** (lower is better) |   |   |  
`PutItem` total seconds (500 items) | 1.5015 | 1.5930 | 1.0562
`BatchWriteItem (25)` total seconds (12k5 items) | 2.5074 | 2.5285 | 2.0132

| DDB Streams | `latest` | Fix & Hypercorn | Fix & Twisted
| --- | --- | --- | --- |
| **Throughput** (higher is better) |  |  |  |
| `PutItem` throughput item/s | 123.14 | 166.73 | 204.27 |
| `BatchWriteItem (25)` throughput item/s | 171.77 | 2161.57 | 2466.23 |
| **Total time** (lower is better) |  |  |  |
| `PutItem` total seconds (500 items) | 4.0604 | 2.9988 | 2.4478 |
| `BatchWriteItem (25)` total seconds (12k5 items) | 72.7713 | 5.7828 | 5.0685 |

| Kinesis | `latest` | Fix & Hypercorn | Fix & Twisted
| --- | --- | --- | --- |
| **Throughput** (higher is better) |  |  |  |
| `PutItem` throughput item/s | 122.19 | 169.20 | 205.97 |
| `BatchWriteItem (25)` throughput item/s | 171.51 | 2534.15 | 2920.69 |
| **Total time** (lower is better) |  |  |  |
| `PutItem` total seconds (500 items) | 4.0920 | 2.9551 |  2.4275 |
| `BatchWriteItem (25)` total seconds (12k5 items) | 72.8815 | 4.9326 | 4.2798 |

Here, we can see that Twisted does make a difference, already in the baseline: **+50%** for `PutItem` and `+20%` for `BatchWriteItem`. It's not going to be as much for the stream ones, same as before, because so much more is happening.

Analyzing the fix, we can now see the fix improved `PutItem` throughput by around **30%**, and Twisted gives it almost an additional **20%** from that. But we're still 3x times slower than without streams.

For `BatchWriteItem`, which was the main concern, we went from a total of 72 seconds to ~5s on average, so almost a **15x** speed and throughput improvement. This is easily explainable from the fetch and batch. Still 2x slower than the baseline, but acceptable. As a funny note, if doing only of those and not the other, we would go from 72 to around 40 only. Only both together allow to relieve the pressure on the Gateway and improve it as much as it does.

### CODE REVIEW

**Review by viren-nadkarni — APPROVED** [review]
This is excellent, great work @bentsku !

File: localstack/services/dynamodb/utils.py [hunk 1]
```diff
@@ -120,15 +128,26 @@ def invalidate_table_schema(cls, table_name: str, account_id: str, region_name:
 
 class ItemFinder:
     @staticmethod
-    def find_existing_item(
-        put_item: Dict, table_name: str, account_id: str, region_name: str
-    ) -> Optional[Dict]:
-        from localstack.services.dynamodb.provider import ValidationException
-
+    def get_ddb_local_client(account_id: str, region_name: str, endpoint_url: str):
+        ddb_local_access_key = f"{account_id}{region_name}".replace("-", "")
```
[the remark this answers is not in the document]
Comment (bentsku) [reply localstack/services/dynamodb/utils.py]: 👍 done! I've pulled out the function and put it in `utils.py`.
